File: src/main/java/de/zalando/zally/rules/SnakeCaseForQueryParamsRule.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.Swagger;
import io.swagger.models.parameters.Parameter;
import io.swagger.models.parameters.QueryParameter;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Lint for snake case for query params
 */
class SnakeCaseForQueryParamsRule {

    private static String SNAKE_CASE_PATTERN = "^[a-z0-9_]*$";

    List<Violation> validate(Swagger swagger) {
        if (swagger == null) return new ArrayList<>();

        List<Violation> violations = new ArrayList<>();
        swagger.getPaths().forEach((path, pathObject) -> {
            pathObject.getOperationMap().forEach((httpMethod, operation) -> {
                operation.getParameters().forEach(param -> {
                    if (param instanceof QueryParameter) {
                        if (!isSnakeCase(param.getName())) {
                            violations.add(getViolation());
                        }
                    }
                });
            });
        });
        return violations;
    }

    private boolean isSnakeCase(String input) {
        Pattern pattern = Pattern.compile(SNAKE_CASE_PATTERN);
        Matcher matcher = pattern.matcher(input);
        return matcher.matches();
    }

    private Violation getViolation() {
        return new Violation(
            "Must: Use snake_case (never camelCase) for Query Parameters",
            "Must: Use snake_case (never camelCase) for Query Parameters",
            ViolationType.MUST,
            "http://zalando.github.io/restful-api-guidelines/naming/Naming.html#must-use-snakecase-never-camelcase-for-query-parameters"
        );
    }
}

File: server/src/main/java/de/zalando/zally/rule/SnakeCaseForQueryParamsRule.kt
package de.zalando.zally.rule

import de.zalando.zally.dto.Violation
import de.zalando.zally.dto.ViolationType
import de.zalando.zally.util.PatternUtil
import io.swagger.models.Swagger
import io.swagger.models.parameters.QueryParameter
import org.springframework.stereotype.Component

/**
 * Lint for snake case for query params
 */
@Component
class SnakeCaseForQueryParamsRule : AbstractRule() {
    override val title = "Use snake_case (never camelCase) for Query Parameters"
    override val url = "http://zalando.github.io/restful-api-guidelines/naming/Naming.html" +
        "#must-use-snakecase-never-camelcase-for-query-parameters"
    override val violationType = ViolationType.MUST
    override val code = "M011"

    override fun validate(swagger: Swagger): Violation? {
        val result = swagger.paths.orEmpty().flatMap { (path, pathObject) ->
            pathObject.operationMap.orEmpty().flatMap { (verb, operation) ->
                val badParams = operation.parameters.filter { it is QueryParameter && !PatternUtil.isSnakeCase(it.name) }
                if (badParams.isNotEmpty()) listOf("$path $verb" to badParams) else emptyList()
            }
        }
        return if (result.isNotEmpty()) {
            val (paths, params) = result.unzip()
            val description = "Parameters that are not in snake_case: " + params.flatten().map { it.name }.toSet().joinToString(",")
            Violation(this, title, description, violationType, url, paths)
        } else null
    }
}
